### PR TITLE
feat: save a JSON file of QC metrics

### PR DESCRIPTION
**Pull request by ivadym**
### This PR:

Creates a QC pydantic model and dumps the summarised metrics into a JSON file. The previous QC workflow has not been deleted.

Added:

Pydantic models that shape the QC metrics JSON generation.
qc_metrics_summary.json is saved in the analysis output directory.

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Review by ashwini06 — COMMENTED** [review]
Well done 🥇 nice implementation! 

Two points, I am thinking related to PR is : 
1) having two similar functions `get_qc_metrics` and `get_qc_metrics_json`. Can we merge this into one and use the latest  `get_qc_metrics_json` in https://github.com/Clinical-Genomics/BALSAMIC/blob/37a1f86aaa5120c6f1b5196c5e97952ddd867b12/BALSAMIC/commands/report/deliver.py#L175 as well?  - @hassanfa 

2) Is this PR only about retrieving metrics or also checking if the qc metrics is passed/failed?
according to our defined thresholds like here: https://github.com/Clinical-Genomics/BALSAMIC/blob/05646bc2f94b1c17eb571489f444f4250e929217/BALSAMIC/utils/constants.py#L131

**Comment by ivadym** [comment]
Thanks for your comments @ashwini06! 

1. Yes, these two functions should be merged. I was planning to do that in another PR (along with some more changes to deliver), but maybe it's better to directly remove `get_qc_metrics()` here, so I will do that.

2. And this PR is only about retrieving metrics and saving them to analysis/qc. The validation part will be addressed in another PR and the `qc_check()` workflow removed.

**Comment by ashwini06** [comment]
> Thanks for your comments @ashwini06!
> 
> 1. Yes, these two functions should be merged. I was planning to do that in another PR (along with some more changes to deliver), but maybe it's better to directly remove `get_qc_metrics()` here, so I will do that.
> 2. And this PR is only about retrieving metrics and saving them to analysis/qc. The validation part will be addressed in another PR and the `qc_check()` workflow removed.

Good, these points are already considered. Then it's better we can merge this PR and take these points as new issues.

**Comment by ashwini06** [comment]
> Kudos, SonarCloud Quality Gate passed!    Quality Gate passed
> 
> [Bug](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=Clinical-Genomics_BALSAMIC&pullRequest=723&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=Clinical-Genomics_BALSAMIC&pullRequest=723) No Coverage information
> [](https://sonarcloud.io/component_measures?id=Clinical-Genomics_BALSAMIC&pullRequest=723&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Clinical-Genomics_BALSAMIC&pullRequest=723&metric=new_duplicated_lines_density&view=list)

@ivadym: Sonarcloud complains about code smell; check if you can reduce the complexity of long `for loop` by splitting some lines into functions, if not possible we can ignore it.

File: BALSAMIC/utils/models.py [hunk 1]
```diff
@@ -687,3 +689,69 @@ class BalsamicWorkflowConfig(BaseModel):
...
+    """Defines the quality control metrics model associated with a specific analysis file
+
+    Attributes:
+        file_name : string (required); quality control analysis file name
+        sequencing_type : List (required); string literal [targeted, wgs]
+        metrics : List (required); quality control metric names
+    """
+
+    file_name: str
+    sequencing_type: List[str]
```
Comment (hassanfa) [inline BALSAMIC/utils/models.py:704]: Please add a validator for this. Something like:

```python
    @validator("sequencing_type", check_fields=False)
    def sequencing_type_literal(cls, value) -> str:
        "Validate sequencing type"
        assert set(value).issubset(
            set(SEQUENCING_TYPE)
        ), f"{value} is not not a valid sequencing type."
        return value
```

File: BALSAMIC/utils/qc_metrics.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 import json
 import os
 
+from BALSAMIC.utils.constants import METRICS
+from BALSAMIC.utils.models import QCCheckModel
+
 MULTIPLE_JSON = [
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/utils/qc_metrics.py:7]: agreed, this part can be removed in this PR.

File: BALSAMIC/utils/models.py [hunk 3]
```diff
@@ -687,3 +689,69 @@ class BalsamicWorkflowConfig(BaseModel):
...
+    metrics: List[str]
+
+
+class QCCheckModel(BaseModel):
```
Comment (hassanfa) [inline BALSAMIC/utils/models.py:708]: Not to be picky, but this doesn't really check it, does it? Mostly it is about its name. It should reflect what it is doing.

My suggestions would be something like: QCExtractionModel

File: tests/utils/test_models.py [hunk 4]
```diff
@@ -392,3 +394,118 @@ def test_params_vep():
 
     # THEN assert values
     assert test_vep_built.vep_filters == "all defaults params"
+
```
Comment (hassanfa) [inline tests/utils/test_models.py:397]: Looks like there are some fixtures here. You should create them in `tests/utils/` and move them there. If it is too much work, you can ignore this comment.

**Review by hassanfa — CHANGES_REQUESTED** [review]
Minor comment on the validator.

Great work ⭐

**Comment by ivadym** [comment]
Updated Dockerfile & container pushed to [clinicalgenomics/balsamic:develop-balsamic_container](https://hub.docker.com/layers/clinicalgenomics/balsamic/develop-balsamic_container/images/sha256-a58a875b2398365d4bd4de99d367785c199275ff72b8deb0cf8d55808095a8f6?context=repo).

File: BALSAMIC/utils/qc_metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/utils/qc_metrics.py]: Addressed in 03b2457. I removed that part and updated `deliver.py` and `qc_report.py` to behave the same way (although they'll be changed in the future I suppose).

File: BALSAMIC/utils/rule.py [hunk 5]
```diff
@@ -103,6 +103,15 @@ def get_variant_callers(
     return list(valid_variant_callers)
 
 
+def get_sequencing_type(config):
+    """
+    input: sample config file from BALSAMIC
+    output: sequencing type string]
```
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/utils/rule.py]: Typo corrected in 6ec51db.

File: BALSAMIC/utils/models.py (hunk 1 above)
Comment (ivadym) [reply BALSAMIC/utils/models.py:704]: Added in the 3ce688c commit.

File: BALSAMIC/utils/models.py (hunk 3 above)
Comment (ivadym) [reply BALSAMIC/utils/models.py]: Changed in 6ec51db.

File: tests/utils/test_models.py (hunk 4 above)
Comment (ivadym) [reply tests/utils/test_models.py:397]: I moved it to `tests/conftest.py` , which already has some fixtures defined if that's ok. Adressed in 3ce688c.

**Review by hassanfa — APPROVED** [review]
Well done. ready to merge!

**Comment by ivadym** [comment]
Solved misbehaviour after force-pushing to feat/QC_check branch. New PR: #754.

**Pull request closed without merging**